docs/architecture/adr-020-deterministic-square-construction.md

```diff
@@ -0,0 +1,82 @@
+# ADR 020: Deterministic Square Construction
+
+## Status
+
+Proposed
+
+## Changelog
+
+- 2023/4/13: initial draft
+
+## Context
+
+The current protocol around the construction of a square (ODS) is based around a set of constraints that are enforced during consensus through validation (See `ProcessProposal`). Block proposers are at liberty to choosing not only what transactions are included and in what order but can effectively decide on the amount of padding (i.e. where each blob is located in the square) and the size of the square. This degree of control leaks needless complexity to users with little upside and allows for adverse behaviour.
+
+Earlier designs were incorporated around the notion of interaction between the block proposer and the transaction submitter. A user that wanted to submit a PFB would go to a potential block proposer, provide them with the transaction, the proposer would then reserve a position in the square for the transaction and finally the transaction submitter would sign the transaction with the provided share index. However, Celestia may have 100 potential block proposers which are often hidden from the network. Furthermore, tranasctions often reach a block proposer through a gossip network, severing the ability for the block proposer to directly communicate with the transaction submitter. Lastly, new transactions with greater fees might arrive causing the block proposer to want to shuffle the transactions around in the square. The response to these problems was to come up with "non-interactive defaults" (first mentioned in [ADR006](./adr-006-non-interactive-defaults.md)).
+
+## Decision
+
+Block proposers should only have the freedom to decide transaction inclusion and ordering (Note there are some existing constraints on ordering). Interaction is still permitted (in the sense that a transaction submitter can directly communicate with a block proposer) but not directly supported by the Celestia network. The non-interactive defaults become mandatory.
+
+Square construction is thus to be reduced to the simple deterministic function:
+
+```go
+func ConstructSquare(txs []Tx) []Share
+```
+
+and it's couterpart
+
+```go
+func DeconstructSquare(shares []Share) []Tx
```

Comment:
Note: the return type here `[]Tx` isn't identical to the `[]Tx` parameter in `ConstructSquare(txs []Tx)` because of index wrapping and separation of blobs.

Should this proposed API elaborate on the `[]Tx` type in each function?

Reply:
On second thought, the index wrapped transactions don't contain the blobs but the blob transactions do so `DeconstructSquare` may also want to return a slice of Blobs 🤔